File: SO/04-14/prodcons.swift
// File: ProducerConsumer.swift
import Foundation

let bufferSize = 5
var buffer = [Int]()
let bufferLock = DispatchQueue(label: "bufferLock")
let empty = DispatchSemaphore(value: bufferSize)
let full = DispatchSemaphore(value: 0)

func producer() {
    DispatchQueue.global().async {
        for i in 0..<10 {
            empty.wait()
            bufferLock.sync {
                buffer.append(i)
                print("Producer produced: \(i)")
            }
            full.signal()
            sleep(1)
        }
    }
}

func consumer() {
    DispatchQueue.global().async {
        for _ in 0..<10 {
            full.wait()
            var item: Int = -1
            bufferLock.sync {
                if !buffer.isEmpty {
                    item = buffer.removeFirst()
                }
            }
            print("Consumer consumed: \(item)")
            empty.signal()
            sleep(2)
        }
    }
}

producer()
consumer()

dispatchMain() // Mantem a aplicacao rodando

File: SO/04-14/prodcons.go
// producer_consumer.go
package main

import (
    "fmt"
    "math/rand"
    "time"
)

const bufferSize = 5

func producer(ch chan<- int) {
    for {
        item := rand.Intn(100)
        ch <- item
        fmt.Printf("Producer produced: %d\n", item)
        time.Sleep(time.Second)
    }
}

func consumer(ch <-chan int) {
    for {
        item := <-ch
        fmt.Printf("Consumer consumed: %d\n", item)
        time.Sleep(2 * time.Second)
    }
}

func main() {
    ch := make(chan int, bufferSize)
    go producer(ch)
    go consumer(ch)

    // Mantem o programa rodando indefinidamente.
    select {}
}
